src/main/java/com/swvl/geometry/shapes/Polygon.java

```diff
@@ -0,0 +1,408 @@
+
+        if (shape instanceof LineSegment)
+            return isLineIntersected((LineSegment) shape);
+
+        throw new OperationNotSupportedException("isIntersected operation in Polygon does not support " + shape.getClass());
+    }
+
+    /**
+     * Returns true if the point p lies inside the polygon
+     * <p>
+     * A point is inside the polygon if either count of intersections is odd or
+     * point lies on an edge of polygon.  If none of the conditions is true, then
+     * point lies outside.
+     */
+    private boolean isPointIntersection(Point point) throws OperationNotSupportedException {
+        if (maxX == Double.MIN_VALUE)
+            for (int i = 0; i < points.length - 1; ++i)
+                maxX = Math.max(maxX, points[i].x);
+
+        /* Create a point for line segment from p to infinite */
+        LineSegment infinityLine = new LineSegment(point, new Point(maxX + 1e3, point.y));
+
+        /* Pointer to current edge in polygon */
+        LineSegment edge = new LineSegment();
+
+        /* Count intersections of the above line with sides of polygon */
+        int count = 0;
+
+
+        for (int i = 0; i < points.length - 1; ++i) {
+            edge.init(points[i], points[i + 1]);
+
+            /*
+             * Check if the infinityLine line segment intersects
+             * with the edge line segment
+             */
+            if (edge.isIntersected(infinityLine)) {
+
+                /* check if it lies on segment. If it lies, return true */
+                if (edge.contains(point))
+                    return true;
+
+                count++;
+            }
+        }
+
+        return count % 2 == 1;
+    }
+
+
+    private boolean isRectangleIntersection(Rectangle rect) throws OperationNotSupportedException {
+        Point p1 = new Point(rect.maxPoint.x, rect.minPoint.y); // bottom-right point
+        Point p2 = new Point(rect.minPoint.x, rect.maxPoint.y); // upper-left point
+
+        /* Edges of a rectangle */
+        LineSegment[] rectEdges = new LineSegment[]{
+                new LineSegment(rect.minPoint, p1),
+                new LineSegment(p1, rect.maxPoint),
+                new LineSegment(rect.maxPoint, p2),
+                new LineSegment(p2, rect.minPoint)
+        };
+
+        Point[] rectPoints = new Point[]{
+                rect.minPoint,
+                p1,
+                rect.maxPoint,
+                p2
+        };
+
+        /* Iterate over edges for checking intersection of edges without any points intersection */
+        for (LineSegment rectEdge : rectEdges)
+            if (this.isLineIntersected(rectEdge))
+                return true;
+
+
+        /*
+         * Check if rectangle is fully inside the polygon. Thus, no edge intersection
+         * Iterate over rectangle points and check if any point is inside the polygon
+         */
+        for (Point rectPoint : rectPoints)
+            if (this.isPointIntersection(rectPoint))
```

Comment:
Won't work for concave polygons

Reply:
I think it will work, because for a concave polygon, if all rectangle's points are inside polygon, any edge which will go out of polygon must return again inside the polygon. Thus, the edge will intersect an even number of polygon's edges.